pandas/core/indexes/base.py

```diff
@@ -2847,8 +2847,32 @@ def get_indexer_for(self, target, **kwargs):
 
     @Appender(_index_shared_docs['_get_values_from_dict'])
     def _get_values_from_dict(self, data):
-        return lib.fast_multiget(data, self.values,
-                                 default=np.nan)
+        # MultiIndex doesn't support isna()
+        if not data or (not hasattr(self, 'levels') and
+                        not isna(self.values).any()):
+            # lib.fast_multiget doesn't support arbitrary dict subclasses
+            if data.__class__ is not dict:
+                data = dict(data)
+            return lib.fast_multiget(data, self.values,
+                                     default=np.nan)
+        else:
+            # Looking for NaN in dict doesn't work ({np.nan : 1}[float('nan')]
+            # raises KeyError), so we iterate the entire dict, and align
+            keys, values = zip(*compat.iteritems(data))
+            keys = Index(keys)
+            values = np.array(values, dtype='object')
+            if not self.equals(keys):
+                locs = self.get_indexer(keys)
+                order = - np.ones(len(self), dtype=int)
+                order[locs] = np.arange(len(keys))
+                values = values[order]
```

Comment:
Isn't this just a `reindex` ?

Reply:
Yes, I just wanted to avoid creating a ``Series`` object, particularly so inside an ``Index`` method... but I can do it if you prefer.

Reply:
Since this method is only used in Series.init, I would then try to refactor to do the reindex there, that would eliminate much of this custom code?

Reply:
Yes... (that would basically be my previous version of this PR, except that I would move code from ``Series.__init__`` to ``Series._init_from_dict``). Notice however that if we leave ``Index._get_values_from_dict`` as it is... it is broken (doesn't work for NaNs).

Reply:
But your previous PR had a lot of complex code checking for tuples etc (I think that part was the main objection), while this one does not have that any more.

Reply:
OK, then if you're OK with leaving ``Index._get_values_from_dict`` as it is, I will try again.